querybuilder/rule_group.go

```diff
@@ -1,41 +1,45 @@
 package querybuilder
 
+import "github.com/pkg/errors"
+
 const (
 	AND = "AND"
 	OR  = "OR"
 )
 
 type Checker interface {
-	Evaluate(dataset map[string]interface{}) bool
+	Evaluate(dataset map[string]interface{}) (bool, error)
 }
 
 type RuleGroup struct {
 	Condition interface{}
 	Rules     interface{}
 }
 
-func (rg *RuleGroup) Evaluate(dataset map[string]interface{}) bool {
+func (rg *RuleGroup) Evaluate(dataset map[string]interface{}) (bool, error) {
 	rules := rg.Rules.([]interface{})
-	var res bool
+
 	switch rg.Condition.(string) {
 	case AND:
 		for _, r := range rules {
-			res = rg.getChecker(r.(map[string]interface{})).Evaluate(dataset)
-			if !res {
-				return false
+			isCorrectInput, err := rg.getChecker(r.(map[string]interface{})).Evaluate(dataset)
+			if err != nil || !isCorrectInput {
+				return isCorrectInput, err
 			}
 		}
-		return true
+		return true, nil
+
 	case OR:
 		for _, r := range rules {
-			res = rg.getChecker(r.(map[string]interface{})).Evaluate(dataset)
-			if res {
-				return true
+			isCorrectInput, err := rg.getChecker(r.(map[string]interface{})).Evaluate(dataset)
+			if err != nil || isCorrectInput {
+				return isCorrectInput, err
```

Comment:
```suggestion
			ok, err := rg.getChecker(r.(map[string]interface{})).Evaluate(dataset)
			if err != nil {
				return false, err
```

Reply:
Two reasons why I chose this approach:-

1) We know that `and` fails when one of the comparisons fails. So, we can reduce the number of comparisons by stop comparing as soon as we encounter a `false`. Similarly `or` passes when one comparison passes. We can stop comparing as soon as we encounter a `false`.

2) We don't store the values in an array or something to finally compute the result. What I mean is say the condition is `AND` and the first check gives `false` and the second gives `true`. We overwrite the same variable that stores that result. So finally the result would be that of the final comparison which in this case isn't right (will return `true` instead of `false`.

Reply:
Ahh, my bad I didn't see the loop before it. So the thing I wanted to say was, 
- If there is a `non-nil` err, we always want to return `false, err`, irrespective of the `ok` because we don't know what `ok` has
- Else we can return `ok` based on some conditions 

Now to do this you either have an option to add nested `if` cases or you can just add it one after the other.

Reply:
Yeah okay.